jsonschema/_types.py

```diff
@@ -0,0 +1,206 @@
+import numbers
+
+import attr
+import pyrsistent
+
+from jsonschema.compat import str_types, int_types, iteritems
+from jsonschema.exceptions import UndefinedTypeCheck
+
+
+def is_array(instance):
+    return isinstance(instance, list)
+
+
+def is_bool(instance):
+    return isinstance(instance, bool)
+
+
+def is_integer(instance):
+    # bool inherits from int, so ensure bools aren't reported as ints
+    if isinstance(instance, bool):
+        return False
+    return isinstance(instance, int_types)
+
+
+def is_null(instance):
+    return instance is None
+
+
+def is_number(instance):
+    # bool inherits from int, so ensure bools aren't reported as ints
+    if isinstance(instance, bool):
+        return False
+    return isinstance(instance, numbers.Number)
+
+
+def is_object(instance):
+    return isinstance(instance, dict)
+
+
+def is_string(instance):
+    return isinstance(instance, str_types)
+
+
+def is_any(instance):
+    return True
+
+
+@attr.s(frozen=True)
+class TypeChecker(object):
+    """
+    A ``type`` property checker.
+
+    A :class:`TypeChecker` performs type checking for an instance of
+    :class:`Validator`. Type checks to perform are set using
+    :meth:`TypeChecker.redefine` or :meth:`TypeChecker.redefine_many` and
+    removed via :meth:`TypeChecker.remove` or
+    :meth:`TypeChecker.remove_many`. Each of these return a new
+    :class:`TypeChecker` object.
+
+    Arguments:
+
+        type_checkers (pyrsistent.pmap):
+
+            It is recommend to set type checkers through
```

Comment:
Hm, is there a reason to recommend that over the param to `__init__`? If there is, I think it'd be better to remove the possibility entirely and stick to just that. (But if we can safely support both, seems fine to do so and to drop the recommendation).

Reply:
Yeah, this is kind-of an awkwardness with the attrs `frozen` class. In order for `evolve` to work, the attribute must be able to be set via `init`. This probably means exposing the underlying type (i.e. `pyrsistent.pmap`), which doesn't seem right.

I suppose that we could have init take in a standard dictionary instead, which is then converted to the pmap by the `TypeChecker` (and then a little juggling in the update methods). This doesn't feel massively clean either but is perhaps the preferred solution.

Thoughts? Sorry this didn't come up earlier, I overlooked this before as the init is generated by attrs.

Reply:
Yeah so this is the juggle to make it work:

```
    _type_checkers = attr.ib(default=())

    def __attrs_post_init__(self):
        object.__setattr__(self, "_type_checkers",
                           pyrsistent.pmap(dict(self._type_checkers)))
```
As it happens, this is the same as the attrs issue you open a few months back: https://github.com/python-attrs/attrs/issues/207

I think this is probably the right solution, however. It means that checkers can be defined on init without users needing to supply a `pmap`. Given this, we could then drop the `redefine_many` method (since the Draft3 `TypeChecker` initialisation was the main internal use of this) but I'm inclined to keep it around.

Reply:
What's making you reluctant to ask for a `pmap` and expect callers to provide one out of curiosity?

(If you did have such a reservation, you can use the `coerce` argument to `attr.ib`, but my personal position happens to be that in 2017, essentially make believe the stdlib types don't exist :)

Reply:
It feels like an implementation detail. To me, the `TypeChecker` should provide an immutable interface but the way in which it does so shouldn't need to be public. Secondarily, it's simple, familiar and more concise for the caller if they can provide a vanilla dict. 

I'm not really too fussy, however.   

(Thanks for the pointer, that's a much cleaner way of doing it - looks like it's `convert` to `attr.ib`)